op-chain-ops: create tool to determine which contract versions are deployed for each chain

0xfuturistic opened the pull request:
Addressing https://github.com/ethereum-optimism/client-pod/issues/405 as part of https://github.com/ethereum-optimism/client-pod/issues/411.

Sample output only passing a Goerli RPC for the `--l1-rpc-url` flag:
```
INFO [01-05|11:41:27.654] OP-Goerli                                l1-chain-id=5 l2-chain-id=420
INFO [01-05|11:41:27.654] Detecting on chain contracts 
INFO [01-05|11:41:29.288] Base Goerli                              l1-chain-id=5 l2-chain-id=84531
INFO [01-05|11:41:29.289] Detecting on chain contracts 
[
  {
    "Name": "OP-Goerli",
    "ChainID": 420,
    "ContractVersions": {
      "L1CrossDomainMessenger": "1.7.0",
      "L1ERC721Bridge": "1.4.0",
      "L1StandardBridge": "1.4.0",
      "L2OutputOracle": "1.6.0",
      "OptimismMintableERC20Factory": "1.6.0",
      "OptimismPortal": "1.10.0",
      "SystemConfig": "1.10.0"
    }
  },
  {
    "Name": "Base Goerli",
    "ChainID": 84531,
    "ContractVersions": {
      "L1CrossDomainMessenger": "1.4.0",
      "L1ERC721Bridge": "1.1.1",
      "L1StandardBridge": "1.1.0",
      "L2OutputOracle": "1.3.0",
      "OptimismMintableERC20Factory": "1.1.0",
      "OptimismPortal": "1.6.0",
      "SystemConfig": "1.3.0"
    }
  }
]
```

Similar to `op-upgrade`, the output file can be specified by a flag. If none is passed, the file is written to stdout (as in the sample output above).

0xfuturistic committed:
op-chain-ops: create op-version-check

forked from op-upgrade

0xfuturistic committed:
op-version-check: drop logic for upgrading contracts

0xfuturistic committed:
op-version-check: add logic for outputting contract versions

0xfuturistic committed:
op-version-check: improve support for checking multiple chains

0xfuturistic committed:
op-version-check: replace legacy upgrade references

0xfuturistic committed:
op-chain-ops: create README

0xfuturistic committed:
op-version-check: improve README

0xfuturistic committed:
op-version-check: drop unnecessary deploy-config flag

0xfuturistic committed:
op-version-check: create op-version-check in Makefile

0xfuturistic committed:
op-version-check: add usage section to README

0xfuturistic committed:
op-chain-ops: add op-version-check to README

0xfuturistic committed:
op-version-check: fix path in README

0xfuturistic committed:
op-version-check: remove unused function toDeployConfigName

0xfuturistic committed:
op-version-check: drop capitalization on error strings

tynes commented on op-chain-ops/cmd/op-version-check/main.go:
```diff
@@ -0,0 +1,191 @@
+				Name:  "chain-ids",
+				Usage: "L2 Chain IDs corresponding to chains to check versions for. Corresponds to all chains if empty",
+			},
+			&cli.StringFlag{
```
Can we replace the `chain-ids` and `superchain-target` flags with a `l2-rpc-urls` flag? This will allow the software to work with L2s that are not part of the superchain. The code to parse the superchain target is brittle and using it in 2 places will create a maintenance burden (its copied from `op-upgrade`).

Could use `StringSliceFlag` - https://github.com/urfave/cli/blob/v2.27.1/flag_string_slice.go

Could also make `l1-rpc-urls` into a string slice flag to make the API symmetric and allow for max flexibility with how the infra team wants to run the software

Using the syntax (# of L1):(# of L2), could run it with 1:1 for a custom network, or 1:many for a superchain or many:0 to observe all L1s

trianglesphere replied:
Do we really need to check more than one L2 / L1 per invocation?
I think the most complex we'd want is to have a single L1 and multiple L2s

tynes replied:
Sure but it is minimal to support multiple L1s - if not then the API is not symmetrical which is weird

0xfuturistic replied:
for l2s that are not part of the superchain, how would the addresses and `chainConfig` for checking the versions be obtained? since the chain is not on the superchain, this raises a tricky question @tynes

tynes replied:
For L2s, the addresses are deterministic, ie can always call the same predeploys

The L1 addresses are not deterministic

0xfuturistic replied:
implemented. lmk if it checks out to you @tynes plz

tynes commented on op-chain-ops/cmd/op-version-check/main.go:
```diff
@@ -0,0 +1,191 @@
+	return "", fmt.Errorf("unsupported chain ID %d", chainID)
+}
+
+func writeJSON(outfile string, input interface{}) error {
```
Do you mind looking in `op-service` to see if there is a JSON utility there for writing a file to disk? I believe there is a JSON package, we could begin to deduplicate this write json function that has been already copied into many places

0xfuturistic replied:
ended up using a function from `op-node/cmd/genesis`!

mslipper commented on op-chain-ops/cmd/op-version-check/main.go:
```diff
@@ -0,0 +1,191 @@
+
+	targets := make([]*superchain.ChainConfig, 0)
+	for _, chainConfig := range superchain.OPChains {
+		if chainConfig.Superchain == superchainName && slices.Contains(chainIDs, chainConfig.ChainID) {
```
Need some logging here if a chain ID is filtered out for whatever reason.

mslipper commented:
We also seem to be missing some contracts in the output, e.g. `AddressManager` and `ProxyAdmin`. See output:

```json
[
  {
    "Name": "OP-Mainnet",
    "ChainID": 10,
    "ContractVersions": {
      "L1CrossDomainMessenger": "1.4.0",
      "L1ERC721Bridge": "1.0.0",
      "L1StandardBridge": "1.1.0",
      "L2OutputOracle": "1.3.0",
      "OptimismMintableERC20Factory": "1.1.0",
      "OptimismPortal": "1.6.0",
      "SystemConfig": "1.3.0"
    }
  },
  {
    "Name": "PGN",
    "ChainID": 424,
    "ContractVersions": {
      "L1CrossDomainMessenger": "1.4.0",
      "L1ERC721Bridge": "1.1.1",
      "L1StandardBridge": "1.1.0",
      "L2OutputOracle": "1.3.0",
      "OptimismMintableERC20Factory": "1.1.0",
      "OptimismPortal": "1.6.0",
      "SystemConfig": "1.3.0"
    }
  },
  {
    "Name": "Base",
    "ChainID": 8453,
    "ContractVersions": {
      "L1CrossDomainMessenger": "1.4.0",
      "L1ERC721Bridge": "1.1.1",
      "L1StandardBridge": "1.1.0",
      "L2OutputOracle": "1.3.0",
      "OptimismMintableERC20Factory": "1.1.0",
      "OptimismPortal": "1.7.0",
      "SystemConfig": "1.3.0"
    }
  },
  {
    "Name": "Zora",
    "ChainID": 7777777,
    "ContractVersions": {
      "L1CrossDomainMessenger": "1.4.0",
      "L1ERC721Bridge": "1.1.1",
      "L1StandardBridge": "1.1.0",
      "L2OutputOracle": "1.3.0",
      "OptimismMintableERC20Factory": "1.1.0",
      "OptimismPortal": "1.6.0",
      "SystemConfig": "1.3.0"
    }
  }
]
```

tynes commented:
Regarding logging the address manager and proxy admin, not sure if that info is in `superchain-registry` at the moment. We will need to add it if not and then can update the calls in `op-chain-ops` to query for them. They are not versioned contracts so perhaps a better resulting JSON would be:

```
  {
    "Name": "OP-Mainnet",
    "ChainID": 10,
    "Contracts": {
      "L1CrossDomainMessenger": {
        "version": "1.4.0",
        "address": "0x..."
      },
      "AddressManager": {
        "version": null,
        "address": "0x..."
      },
      ....
    }
  },
```

0xfuturistic committed:
op-node: capitalize writeJSONFile function

this makes it accessible when op-node/cmd/genesis is imported as a package

0xfuturistic committed:
op-chain-ops: use op-node/cmd/genesis for writeJSON

0xfuturistic committed:
op-chain-ops: set name op_node_genesis for import

0xfuturistic committed:
op-chain-ops: use op-node WriteJSONFile in op-upgrade

0xfuturistic committed:
op-chains-ops: move to toSuperchainName to upgrades package

0xfuturistic committed:
op-chain-ops: output all contracts for op-version-check

addresses https://github.com/ethereum-optimism/optimism/pull/8859#issuecomment-1881663953

0xfuturistic committed:
op-chain-ops: add logging if chain ID is skipped by op-version-check

tynes commented on op-chain-ops/cmd/op-upgrade/main.go:
```diff
@@ -19,6 +19,7 @@ import (
 	"github.com/ethereum-optimism/optimism/op-chain-ops/safe"
 	"github.com/ethereum-optimism/optimism/op-chain-ops/upgrades"
 
+	op_node_genesis "github.com/ethereum-optimism/optimism/op-node/cmd/genesis"
```
nit: very non standard to import things from `cmd` directory. `cmd` doesn't contain libraries, it should contain the minimal stuff to make an executable. You should move `WriteJSON` to the json package in `op-service`

0xfuturistic replied:
addressed. thank you!

0xfuturistic committed:
op-service: create jsonutil/write

0xfuturistic committed:
op-chain-ops use jsonutil in op-upgrade, op-version-check

0xfuturistic committed:
op-chain-ops: pass RPC URLs as flags in op-version-check

0xfuturistic committed:
op-chain-ops: refactor of op-version-check

0xfuturistic committed:
op-chain-ops: refactor of op-version-check

0xfuturistic committed:
op-chain-ops: add more logging for op-version-check

0xfuturistic committed:
op-chain-ops: add more logging for op-version-check

tynes commented on op-chain-ops/cmd/op-version-check/main.go:
```diff
@@ -0,0 +1,164 @@
+	l2RPCURLs := ctx.StringSlice("l2-rpc-urls")
+
+	var l2ChainIDs []uint64
+
```
- For each L2 RPC:
  - Call `version()` on each versioned predeploy
- For each L1 RPC:
  - Grab the contract addresses from superchain registry
  - For each contract address:
    - Call `version()`

0xfuturistic commented on op-chain-ops/cmd/op-version-check/main.go:
```diff
@@ -0,0 +1,164 @@
+			addresses, ok := superchain.Addresses[l2ChainID]
+			if !ok {
+				return fmt.Errorf("no addresses for chain ID %d", l2ChainID)
+			}
```
we want to be able to call `version()` in the Predeploys for L2 contracts
